### PR TITLE
Sphinx autoprocess extension

### PR DESCRIPTION
**Pull request by JureZmrzlikar**
No more corrections anticipated.

### CODE REVIEW

File: docs/autoprocess.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+            * Process slug - if the content of directive is a valid process
+              slug only corresponding process is returned. Actually, a list
+              with only this process is returned.
+            * If none of this options, then raise a ValueError
+
+        :param content: content of the directive
+        :type content: list
+        :param categories: set of all categories
+        :type categories: set
+        :param processes: dict of all github_url: processes pairs
```
Comment (dblenkus) [inline docs/autoprocess.py]: `processes` -> `processes_dict`

File: docs/autoprocess.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+            * None - if there is not content, all processes are returned.
+            * Category - if the content of the directive is a string matching
+              an element in categories list, only processes with this
+              category will be returned.
+            * Process slug - if the content of directive is a valid process
+              slug only corresponding process is returned. Actually, a list
+              with only this process is returned.
+            * If none of this options, then raise a ValueError
+
+        :param content: content of the directive
```
Comment (dblenkus) [inline docs/autoprocess.py]: For basic types you can use `:param list content: ...` and omit the second line.

http://www.sphinx-doc.org/en/stable/domains.html#info-field-lists

File: docs/autoprocess.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+        :type content: list
+        :param categories: set of all categories
+        :type categories: set
+        :param processes: dict of all github_url: processes pairs
+        :type processes:dict
+
+        :rtype: list
+        """
+        if not content:
+            return processes
```
Comment (dblenkus) [inline docs/autoprocess.py]: `processes` is not defined. Probably `processes_dict`?

File: docs/autoprocess.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+        else:
+            content = content[0]
```
Comment (dblenkus) [inline docs/autoprocess.py]: What happens with other elements of the list?

File: docs/autoprocess.py [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+        :param categories: set of all categories
+        :type categories: set
+        :param processes: dict of all github_url: processes pairs
+        :type processes:dict
+
+        :rtype: list
+        """
+        if not content:
+            return processes
+        else:
```
Comment (dblenkus) [inline docs/autoprocess.py]: It will be more readable if you omit `else` and remove one level of indentation.

File: docs/autoprocess.py [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+            querry_for_slug = dict(pair for pair in processes_dict.items() if pair[1]['slug'] == content)
+            if len(querry_for_slug.items()) == 1:
+                return querry_for_slug
+
+            elif content in categories:
+                if content == 'uncategorized':
+                    filtered_dict = dict(
+                        pair for pair in processes_dict.items() if 'category' not in pair[1])
+                    return filtered_dict
+                else:
```
Comment (dblenkus) [inline docs/autoprocess.py]: Remove `else`

File: docs/autoprocess.py [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+                    filtered_dict = dict(
+                        pair for pair in processes_dict.items()
+                        if 'category' in pair[1] and pair[1]['category'] == content)
+                    return filtered_dict
+
+            else:
```
Comment (dblenkus) [inline docs/autoprocess.py]: Remove `else`.

File: docs/autoprocess.py [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+                raise ValueError("Invalid argument given to autoprocess directive.")
+
+    def get_all_processes(self, process_dir, github_base_url):
+        """
+        Return a dictionary of all processes in `process_dir`
+
+        Dictionary key is an URL's to process source code on resolwe-bio
+        GitHub page and value is process as it is loaded from yaml file.
```
Comment (dblenkus) [inline docs/autoprocess.py]: Is this linked to GitHub in any way? If not, `source_base_url` would be better.

File: docs/autoprocess.py [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+        for yaml_file in all_process_files:
+            processes_in_file = read_yaml_file(yaml_file)
+            for process in processes_in_file:
+
+                # This section finds the line in file where the
+                # defintion of the process starts. (there are
+                # multiple process definition in some files).
+                i = 1
+                found = False
+                with open(yaml_file) as file_:
```
Comment (dblenkus) [inline docs/autoprocess.py]: You've already read the file in `read_yaml_file` function. Why you have to do it again?

File: docs/autoprocess.py [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+                    for line in file_:
+                        if 'slug: ' + process['slug'] in line:
```
Comment (dblenkus) [inline docs/autoprocess.py]: There is no guarantee that the `slug` attribute will be in the first line of the process.

File: docs/autoprocess.py (hunk 4 above)
Comment (JureZmrzlikar) [inline docs/autoprocess.py]: Currently, they are discarded.

File: docs/autoprocess.py [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+
+        For processes with undefined category field (yes, there are some!),
+        there is uncategorized category.
+
+        :param process_list: list of all processes
+        :type process_list: list
+
+        :rtype: list
+        """
+        categories = ['uncategorized']
```
Comment (dblenkus) [inline docs/autoprocess.py]: categories = set()
    for process in process_list:
        categories.add(process.get('category', 'uncategorized'))
    return list(categories)

File: docs/autoprocess.py (hunk 9 above)
Comment (JureZmrzlikar) [inline docs/autoprocess.py]: With `read_yaml_file` function I directly turn yaml to dictionary. The second opening is juts to determine the starting line of process definition. I don't know how to do it simultaneously.

File: docs/autoprocess.py (hunk 10 above)
Comment (JureZmrzlikar) [inline docs/autoprocess.py]: Indeed, but that was the best idea that i had. Suggestions?

File: docs/autoprocess.py [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+        title = nodes.subtitle()
+        title_text = nodes.Text(name)
+        title += title_text
+        section += title
+
+        top_field_list = nodes.field_list()
+
+        slug = process['slug']
+        version = process['version']
+        description = process['description']
```
Comment (dblenkus) [inline docs/autoprocess.py]: There is a chance that `description` is not given.

    description = process.get('description', '')

File: docs/autoprocess.py [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+        inputs = process['input'] if 'input' in process else []
```
Comment (dblenkus) [inline docs/autoprocess.py]: process.get('input', [])

same for `output`

File: docs/autoprocess.py [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+              representing the hierarchical structure of the document. More
+              importantly, the whole point of method run is to return a
+              list of nodes. Theese nodes are than parsed by docutils
+              package into HTML/LaTeX...
+
+        """
+
+        # Get all processes:
+        all_processes_dict = self.get_all_processes(PROCESS_DIR, GITHUB_BASE_URL)
+        _, all_processes = zip(*all_processes_dict.items())
```
Comment (dblenkus) [inline docs/autoprocess.py]: Maybe `all_processes = zip(*all_processes_dict.items())[1]`?

File: docs/conf.py [hunk 15]
```diff
@@ -5,6 +5,9 @@
 import os
 import shlex
 
+# Add the folder where autoprocess.py extension is located to python path, so sphinx can locate it:
+sys.path.append(os.path.abspath(os.path.dirname(__file__)))
```
Comment (dblenkus) [inline docs/conf.py]: Save `os.path.abspath(os.path.dirname(__file__))` to variable and use here and in `base_dir = ...`

File: docs/autoprocess.py (hunk 9 above)
Comment (dblenkus) [inline docs/autoprocess.py]: Ok, I've figured it out :). Move this part to function, so you can just return `i` when process is found or return `1` at the end.

File: docs/autoprocess.py [hunk 16]
```diff
@@ -0,0 +1,412 @@
...
+            paragraph += nodes.Text(field_body)
+        else:
+            # This is the case when field_body is a complex node:
+            # useful when constructing nested field lists
+            paragraph += field_body
+
+        body += paragraph
+        field += name
+        field += body
+        return field
```
Comment (dblenkus) [inline docs/autoprocess.py]: name = nodes.field_name() + nodes.Text(field_name)

    if isinstance(field_body, six.string_types):
        field_body = nodes.Text(field_body)
    paragraph = nodes.paragraph() + field_body

    body = nodes.field_body() + paragraph

    return nodes.field() + name + body

Something like this looks more readable to me.

File: docs/autoprocess.py [hunk 17]
```diff
@@ -0,0 +1,401 @@
...
+        :return: line where the process definiton starts
+        :rtype: int
+        """
+        i = 1
+        found = False
+        with open(fname) as file_:
+            for line in file_:
+                if 'slug: ' + slug in line:
+                    found = True
+                    break
```
Comment (dblenkus) [inline docs/autoprocess.py]: You don't need `found` anymore. Just `return i` instead of `break`.

File: docs/autoprocess.py [hunk 18]
```diff
@@ -0,0 +1,401 @@
...
+                # Special formating of `value` is needed if name == 'choices'
+                elif name == 'choices':
+                    bullet_list = nodes.bullet_list()
+                    for choice in value:
+                        list_item = nodes.list_item()
+                        paragraph = nodes.paragraph()
+                        label = nodes.Text(choice['label'] + ': ')
+                        val = nodes.literal(choice['value'], choice['value'])
+
+                        paragraph += label
```
Comment (dblenkus) [inline docs/autoprocess.py]: Move `paragraph` definition just before it is used. Or maybe even `paragraph = nodes.paragraph() + label + val`?

File: docs/autoprocess.py [hunk 19]
```diff
@@ -0,0 +1,401 @@
...
+                        paragraph += val
+                        list_item += paragraph
```
Comment (dblenkus) [inline docs/autoprocess.py]: Same for `list_item`

File: docs/autoprocess.py [hunk 20]
```diff
@@ -0,0 +1,401 @@
...
+        :param url/string source_base_url: url to process definition
+        :param dict process: process data as given from yaml.load function
+        :rtype: nodes.field
+        """
+
+        name = process['name']
+
+        # Make process name as section title:
+        section = nodes.section()
+        section['ids'] = [str(uuid.uuid4())]
```
Comment (dblenkus) [inline docs/autoprocess.py]: Is this id visible to the end user? If yes, it should be nice, i.e. use process slug.

**Comment by dblenkus** [comment]
Try not to use `github` in the code if it is not directly link to the GitHub (i.e. GITHUB_BASE_URL is fine). Something like `source_code` or just `source` would be better.

File: docs/autoprocess.py [hunk 21]
```diff
@@ -0,0 +1,401 @@
...
+
+class AutoProcessDirective(Directive):
+    """
+    Directive used to automatically document resolwe-bio process.
+    """
+    has_content = True
+    required_arguments = 0
+    optional_arguments = 0
+    final_argument_whitespace = False
+    option_spec = None
```
Comment (dblenkus) [inline docs/autoprocess.py]: Where are these variables used? Are they required by Sphinx?

File: docs/autoprocess.py (hunk 16 above)
Comment (JureZmrzlikar) [inline docs/autoprocess.py]: Unfortunately it cannot be done this way. The "+" operator for docutils.nodes is same as append operator. Therefore
`name += nodes.Text(field_name)`
is same as 
`name.append(nodes.Text(field_name))`

File: docs/autoprocess.py (hunk 20 above)
Comment (JureZmrzlikar) [inline docs/autoprocess.py]: Not visible to end user, but still, better to put process slug.

File: docs/autoprocess.py (hunk 21 above)
Comment (JureZmrzlikar) [inline docs/autoprocess.py]: Yes, every child of `Directive` has to define these attributes.

File: docs/autoprocess.py [hunk 22]
```diff
@@ -0,0 +1,397 @@
...
+
+        elif content in categories:
+            if content == 'uncategorized':
+                filtered_dict = dict(
+                    pair for pair in processes_dict.items() if 'category' not in pair[1])
+                return filtered_dict
+
+            filtered_dict = dict(
+                pair for pair in processes_dict.items()
+                if 'category' in pair[1] and pair[1]['category'] == content)
```
Comment (dblenkus) [inline docs/autoprocess.py]: `if pair[1].get('category', 'uncategorized') == content` and you can omit if above.

File: docs/autoprocess.py [hunk 23]
```diff
@@ -0,0 +1,397 @@
...
+        Find the starting line of process definition - where slug is defined
+
+        :param path/str fname: Filename with processes
+        :param string slug: slug of the process
+        :return: line where the process definiton starts
+        :rtype: int
+        """
+        i = 1
+        with open(fname) as file_:
+            for line in file_:
```
Comment (dblenkus) [inline docs/autoprocess.py]: `for i, line in enumerate(file_):` and delete `i = 1` and `i += 1`

File: docs/autoprocess.py [hunk 24]
```diff
@@ -0,0 +1,397 @@
...
+
+import six
+import yaml
+
+from docutils import nodes
+from docutils.parsers.rst import Directive
+
+
+PROCESS_DIR = os.path.normpath(os.path.dirname(__file__) + "./../resolwe_bio/processes/")
+SOURCE_BASE_URL = 'https://github.com/genialis/resolwe-bio/blob/master/resolwe_bio/processes/'
```
Comment (dblenkus) [inline docs/autoprocess.py]: Is there an option to pass this values as arguments in `rst` file? This directive will be also used in other packages that contains processes, so `resolwe-bio` shouldn't be hardcoded.

File: docs/autoprocess.py [hunk 25]
```diff
@@ -0,0 +1,397 @@
...
+                # This section finds the line in file where the
+                # defintion of the process starts. (there are
+                # multiple process definition in some files).
+                startline = self.get_process_definition_start(yaml_file, process['slug'])
+
+                # Put together URL to starting line of process definition.
+                cutted_yaml_path = yaml_file.split('/resolwe_bio/processes/')
+                source_base_url = source_base_url + cutted_yaml_path[1] + '#L' + str(startline)
+                chosen_process = [proc for proc in processes_in_file if proc['slug'] == process['slug']]
+                assert len(chosen_process) == 1
```
Comment (dblenkus) [inline docs/autoprocess.py]: What happens if this is not `True`? Add message with the description of the error.

File: docs/autoprocess.py (hunk 24 above)
Comment (JureZmrzlikar) [inline docs/autoprocess.py]: Good point: It actually should be done this way... we need to define new config_value. Will do it.

File: docs/autoprocess.py [hunk 26]
```diff
@@ -0,0 +1,397 @@
...
+        processes_to_document = self.filter_processes(self.content, categories, all_processes_dict)
+
+        if len(processes_to_document.items()) == 1:
+            # Only one process:
+            source_url, process = processes_to_document.items()[0]
+            top_node = self.make_process_node(source_url, process)
+        elif len(processes_to_document.items()) > 1:
+            # A category of processes, make a process category section:
+            process1 = processes_to_document.items()[0][1]
+            category = process1['category'] if 'category' in process1 else "uncategorized"
```
Comment (dblenkus) [inline docs/autoprocess.py]: `process1.get('category', 'uncategorized')`

File: docs/autoprocess.py [hunk 27]
```diff
@@ -0,0 +1,397 @@
...
+        :param set categories: set of all categories
+        :param dict processes_dict: dict of all source_base_url: processes pairs
+        :rtype: list
+        """
+        if not content:
+            return processes_dict
+        content = content[0]
+        # Make a filtered dict of all process that have name == content
+        querry_for_slug = dict(pair for pair in processes_dict.items() if pair[1]['slug'] == content)
+        if len(querry_for_slug.items()) == 1:
```
Comment (dblenkus) [inline docs/autoprocess.py]: You can ommit `.items()`

File: docs/autoprocess.py (hunk 16 above)
Comment (dblenkus) [inline docs/autoprocess.py]: Ok, then just define variables just before they are used. This way it is easier to figure it out where they come from.

File: docs/autoprocess.py (hunk 10 above)
Comment (dblenkus) [inline docs/autoprocess.py]: Now, that I've found out, why this is used for, I think that solution is sufficient. Solution would  be to look for first line of yaml element and return it when you find right slug, but this is overkill for this use-case.

File: docs/autoprocess.py (hunk 4 above)
Comment (dblenkus) [inline docs/autoprocess.py]: Raise `NotImplementedError` if there is more than one element.

File: docs/index.rst [hunk 28]
```diff
@@ -20,6 +20,7 @@ Contents
    intro
    proc
    contributing
+   catalog
 
 Process Catalog
 ===============
```
Comment (mstajdohar) [inline docs/index.rst]: Remove the Process Catalog and Pipeline Examples sections.

**Pull request merged**